java/source/flight.rst

```diff
@@ -0,0 +1,1061 @@
+
+    // Server
+    Location location = Location.forGrpcInsecure("0.0.0.0", 33333);
+    Map<FlightDescriptor, DataInMemory> dataInMemory = new HashMap<>();
+    Map<String, DataInMemory> mapPojoFlightDataInMemory = new HashMap<>();
+    List<ArrowRecordBatch> listArrowRecordBatch = new ArrayList<>();
+    try (RootAllocator allocator = new RootAllocator(Long.MAX_VALUE)){
+        FlightServer flightServer = FlightServer.builder(allocator, location, new NoOpFlightProducer(){
+            @Override
+            public Runnable acceptPut(CallContext context, FlightStream flightStream, StreamListener<PutResult> ackStream) {
+                return () -> {
+                    long rows = 0;
+                    while (flightStream.next()) {
+                        VectorUnloader unloader = new VectorUnloader(flightStream.getRoot());
+                        try (final ArrowRecordBatch arb = unloader.getRecordBatch()) {
+                            // Retain data information
+                            listArrowRecordBatch.add(arb);
+                            rows = rows + flightStream.getRoot().getRowCount();
+                        }
+                    }
+                    long finalRows = rows;
+                    DataInMemory pojoFlightDataInMemory = new DataInMemory(listArrowRecordBatch, flightStream.getSchema(), finalRows);
+                    dataInMemory.put(flightStream.getDescriptor(), pojoFlightDataInMemory);
+                    ackStream.onCompleted();
+                };
+            }
+
+            @Override
+            public void getStream(CallContext context, Ticket ticket, ServerStreamListener listener) {
+                FlightDescriptor flightDescriptor = FlightDescriptor.path(new String(ticket.getBytes(), StandardCharsets.UTF_8)); // Recover data for key configured
+                if(dataInMemory.containsKey(flightDescriptor)){
+                    VectorSchemaRoot vectorSchemaRoot = VectorSchemaRoot.create(dataInMemory.get(flightDescriptor).getSchema(), allocator);
+                    listener.start(vectorSchemaRoot);
+                    for(ArrowRecordBatch arrowRecordBatch : dataInMemory.get(flightDescriptor).getListArrowRecordBatch()){
+                        vectorSchemaRoot.allocateNew();
+                        VectorLoader loader = new VectorLoader(vectorSchemaRoot);
+                        loader.load(arrowRecordBatch.cloneWithTransfer(allocator));
+                        listener.putNext();
+                    }
+                    vectorSchemaRoot.clear();
+                    listener.completed();
+                }
+            }
+
+            @Override
+            public void doAction(CallContext context, Action action, StreamListener<Result> listener) {
+                FlightDescriptor flightDescriptor = FlightDescriptor.path(new String(action.getBody(), StandardCharsets.UTF_8)); // For recover data for key configured
+                if(dataInMemory.containsKey(flightDescriptor)) {
```

Comment:
We might want to use ConcurrentHashMap and show 'proper' handling of multithreading.

Reply:
Changed

Reply:
Wouldn't the right way to handle this be remove and catch the error? Since another thread may remove it concurrently, we shouldn't be checking before removing.

Reply:
Reviewing ConcurrentHashMap it is thread safe. Is it needed to add that validation?

Reply:
Individual operations are thread safe. Combinations of operations are not thread safe. This is a time-of-check-vs-time-of-use-error. If two threads call this simultaneously, both will check that the key exists, and then both try to remove it, and one will get an error. Instead, they should both just assume the key exists, remove it, and swallow the error if it happens.

Reply:
Thanks for the clarification. Updated